### ISSUE TITLE
minio client config host add <alias> not working properly

### ISSUE THREAD

**Issue by kmala** [minio/mc#1621]
minio client config host add <alias> is not working properly.It is properly updating the config.json but other client commands aren't working.
Temporary fix: Delete the ~/.mc folder and try adding again

**Comment by harshavardhana** [comment]
> minio client config host add is not working properly.It is properly updating the config.json but other client commands aren't working.
> Temporary fix: Delete the ~/.mc folder and try adding again

Easy to reproduce as mentioned related to 'loadMcConfig()' and caching.

**Comment by sreeram-boyapati** [comment]
I would like to take this up, looks like an easy one to fix. When I wrote tests for mc config host. I was able to list, add and remove hosts. Is the issue still open?

**Comment by harshavardhana** [comment]
> I would like to take this up, looks like an easy one to fix. When I wrote tests for mc config host. I was able to list, add and remove hosts. Is the issue still open?

It might not be fixed just yet, it might be still reproducible if you have a custom folder. Basically if you have a "~/.mc" directory then "--config-dir" option adds the host but fails to list or complains about alias being missing. 

To reproduce you should already have a ".mc" directory.

**Comment by sreeram-boyapati** [comment]
Will try to fix it. The Issue is still reproducible.

**Comment by harshavardhana** [comment]
> Will try to fix it. The Issue is still reproducible.

It is a little tricky! , ping us if you have more questions @sreeram-boyapati :-)

**Comment by harshavardhana** [comment]
> Will try to fix it. The Issue is still reproducible.

Thanks a lot! :cake:

**Comment by sreeram-boyapati** [comment]
@harshavardhana 
```go
func getMcConfigPath() (string, *probe.Error) {
	if mcCustomConfigDir == "" {
		fmt.Println("Custom Config is NUll")
	}
	if mcCustomConfigDir != "" {
		return filepath.Join(mcCustomConfigDir, globalMCConfigFile), nil
	}
	dir, err := getMcConfigDir()
	if err != nil {
		return "", err.Trace()
	}
	return filepath.Join(dir, globalMCConfigFile), nil
}
```
mcCustomConfigDir is always empty

**Comment by harshavardhana** [comment]
> func getMcConfigPath() (string, *probe.Error) {
>     if mcCustomConfigDir == "" {
>         fmt.Println("Custom Config is NUll")
>     }
>     if mcCustomConfigDir != "" {
>         return filepath.Join(mcCustomConfigDir, globalMCConfigFile), nil
>     }
>     dir, err := getMcConfigDir()
>     if err != nil {
>         return "", err.Trace()
>     }
>     return filepath.Join(dir, globalMCConfigFile), nil
> }

It is because of the closure that is used in loadMcConfig(), causes a race. That is the bug that we need to fix here.

**Comment by harshavardhana** [comment]
This is fixed now by #1666
